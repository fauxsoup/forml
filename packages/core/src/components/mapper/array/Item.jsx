import debug from 'debug';
import React, { forwardRef, useMemo } from 'react';
import { Draggable } from 'react-beautiful-dnd';

import { useDecorator, useLocalizer, useArrayActions } from '@forml/hooks';
import { ARRAY_PLACEHOLDER } from '../../../constants';
import { clone, traverseForm } from '../../../util';
import { SchemaField } from '../../schema-field';

const log = debug('forml:core:array:item');

export const Item = forwardRef(function Item(props, ref) {
    const { form, parent, forms, index, disabled } = props;
    const dragDrop = useMemo(
        () => ('dragDrop' in form ? form.dragDrop : true),
        [parent]
    );

    const Component = useMemo(() => {
        if (dragDrop) {
            return DraggableArrayItem;
        } else {
            return NormalArrayItem;
        }
    }, [dragDrop]);

    const fields = useMemo(
        () =>
            forms.map((template, subFormIndex) => {
                if (!template) return;
                const form = copyWithIndex(template, index);

                form.titleFun =
                    'titleFun' in form ? form.titleFun : parent.titleFun;
                form.readonly =
                    'readonly' in form ? form.readonly : disabled;

                return (
                    <SchemaField
                        key={subFormIndex}
                        form={form}
                        schema={form.schema}
                        parent={parent}
                        onChange={props.onChange}
                    />
                );
            }),
        [forms, index, parent.titleFun, disabled]
    );

    return (
        <Component {...props} ref={ref}>
            {fields}
        </Component>
    );
});

export default Item;

const NormalArrayItem = forwardRef(function NormalArrayItem(props, ref) {
    const { form, array, value, index, onChange, dragHandleProps, draggableProps } =
        props;
    const { readonly: disabled } = form;
    const deco = useDecorator();
    const localizer = useLocalizer();

    const title = useMemo(() => {
        let title = form.title;
        if (form.titleFun) {
            title = form.titleFun(value);
        }

        return localizer.getLocalizedString(title);
    }, [form, value, localizer]);

    const arrayActions = useArrayActions(form.key);
    const actions = useMemo(function() {
        return {
            destroy: function() {
                const nextModel = arrayActions.removeArray(index);
                onChange(new Event('change', { bubbles: true }), nextModel);
                return nextModel;
            },
            moveUp: function() {
                const nextModel = arrayActions.moveArray(index, index - 1)
                onChange(new Event('change', { bubbles: true }), nextModel);
                return nextModel;
            },
            moveDown: function() {
                const nextModel = arrayActions.moveArray(index, index + 1)
                onChange(new Event('change', { bubbles: true }), nextModel);
                return nextModel;
            },
        };
    }, [
        arrayActions,
        form.key,
        index,
        onChange
    ]);

    return (
        <deco.Arrays.Item
            disabled={disabled}
            title={title}
            index={index}
            form={form}
            dragHandleProps={dragHandleProps}
            draggableProps={draggableProps}
            {...actions}
            ref={ref}
        >
            {props.children}
        </deco.Arrays.Item>
    );
});

function DraggableItemFactory(props, ref) {
    return function DraggableItem(provided) {
        const injectRef = (e) => {
            provided.innerRef(e);
            if (ref) ref(e);
        };
        return (
            <NormalArrayItem {...props} {...provided} ref={injectRef}>
                {props.children}
            </NormalArrayItem>
        );
    };
}

const DraggableArrayItem = forwardRef(function DraggableArrayItem(props, ref) {
    return (
        <Draggable draggableId={props.id} index={props.index}>
            {DraggableItemFactory(props, ref)}
        </Draggable>
    );
});

function copyWithIndex(form, index) {
    const copy = clone(form);
    copy.arrayIndex = index;
    traverseForm(copy, setIndex(index));
    return copy;
}

function setIndex(index) {
    return function(form) {
        if (form.key) {
            form.key[form.key.indexOf(ARRAY_PLACEHOLDER)] = index;
        }
    };
}
