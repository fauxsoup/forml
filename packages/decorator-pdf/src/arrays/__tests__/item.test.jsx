import Item from "../item";
import Context from "@forml/context";
import React from "react";
import ReactPDF from "@react-pdf/renderer";
import { render } from "@testing-library/react";
import * as decorator from "../../";

describe("renders", function() {
  let form;
  let schema;
  let title = "title";
  let description = "description";
  let styles;
  let itemRoot;

  beforeEach(function() {
    schema = { type: "string" };
    form = { type: "item", items: [{ key: [] }] };
    itemRoot = jest.fn(() => ({ border: "1px solid red" }));
    styles = {
      get itemRoot() {
        return itemRoot();
      },
    };
  });

  describe("with form options", function() {
    let fields = {
      wrap: [true, false],
    };

    Object.keys(fields).forEach(function(field) {
      fields[field].forEach(function(value) {
        describe(`${field}`, function() {
          test(`${value}`, function() {
            form = { ...form, [field]: value };
            const { container } = render(
              <ReactPDF.Page>
                <Context.Provider value={{ decorator }}>
                  <Item form={form} title={title} description={description} />
                </Context.Provider>
              </ReactPDF.Page>
            );

            expect(container).toMatchSnapshot();
          });
        });
      });
    });
  });

  test("with title and description", function() {
    const { container } = render(
      <ReactPDF.Page>
        <Context.Provider value={{ decorator }}>
          <Item form={form} title={title} description={description} />
        </Context.Provider>
      </ReactPDF.Page>
    );

    expect(container).toMatchSnapshot();
  });

  test("with title and no description", function() {
    const { container } = render(
      <ReactPDF.Page>
        <Context.Provider value={{ decorator }}>
          <Item form={form} title={title} />
        </Context.Provider>
      </ReactPDF.Page>
    );

    expect(container).toMatchSnapshot();
  });

  test("with description and no title", function() {
    const { container } = render(
      <ReactPDF.Page>
        <Context.Provider value={{ decorator }}>
          <Item form={form} description={title} />
        </Context.Provider>
      </ReactPDF.Page>
    );

    expect(container).toMatchSnapshot();
  });

  test("with no title or description", function() {
    const { container } = render(
      <ReactPDF.Page>
        <Context.Provider value={{ decorator }}>
          <Item form={form} />
        </Context.Provider>
      </ReactPDF.Page>
    );

    expect(container).toMatchSnapshot();
  });

  test("with relevant formStyles", function() {
    form = { ...form, styles };

    const { container } = render(
      <ReactPDF.Page>
        <Context.Provider value={{ decorator }}>
          <Item form={form} />
        </Context.Provider>
      </ReactPDF.Page>
    );

    expect(container).toMatchSnapshot();
    expect(itemRoot).toHaveBeenCalled();
  });
});
