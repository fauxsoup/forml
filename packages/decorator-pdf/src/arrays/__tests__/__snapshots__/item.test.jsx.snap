// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with description and no title 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;

exports[`renders with form options wrap false 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;

exports[`renders with form options wrap true 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;

exports[`renders with no title or description 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;

exports[`renders with relevant formStyles 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column; border: 1px solid red;"
    />
  </page>
</div>
`;

exports[`renders with title and description 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;

exports[`renders with title and no description 1`] = `
<div>
  <page>
    <view
      style="display: flex; flex: 1; flex-direction: column;"
    />
  </page>
</div>
`;
